File: services/mcsrvstat.py
"""Публичный статус серверов Minecraft.

Три источника статуса (в порядке попыток):
  1. REST API mcsrvstat.us — быстрый путь для обычных серверов;
  2. modern Server List Ping (SLP) на «правильном» порту — порт из
     server_meta / переданного параметра / ответа mcsrvstat API / 25565;
  3. legacy ping (0xFE 0x01) — Aternos-серверы часто отвечают только на него
     (modern SLP молчит), поэтому это страховка для них.

Никогда не бросает исключений: недоступность источника = offline-словарь.
"""

from __future__ import annotations

import asyncio
import json
import logging
import struct
from typing import Dict, List, Optional

import aiohttp

logger = logging.getLogger(__name__)

API_URL = "https://api.mcsrvstat.us/3/{ip}"
LEGACY_PING_TIMEOUT = 5.0  # сек
MODERN_SLP_TIMEOUT = 4.0   # сек
REQUEST_TIMEOUT = 7.0      # сек, для mcsrvstat.us
HTTP_UA = (
    "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36"
    " (KHTML, like Gecko) Chrome/126.0 Safari/537.36"
)


def _empty_payload(server_ip: str) -> dict:
    """Стандартизированный ответ при недоступности сервера."""
    return {
        "ip": server_ip,
        "is_online": False,
        "players_online": 0,
        "players_max": 0,
        "player_names": [],
        "version": "Неизвестно",
    }


def _parse_payload(server_ip: str, data: dict) -> dict:
    """Нормализует JSON mcsrvstat в стандартизированный словарь."""
    players = data.get("players") or {}
    raw_names: List[dict] = players.get("list") or []
    return {
        "ip": server_ip,
        "is_online": bool(data.get("online", False)),
        "players_online": int(players.get("online") or 0),
        "players_max": int(players.get("max") or 0),
        "player_names": [
            str(item["name"]) for item in raw_names if isinstance(item, dict) and item.get("name")
        ],
        "version": str(data.get("version") or "") or "Неизвестно",
        "port": int(data["port"]) if data.get("port") else None,
    }


async def _modern_slp(host: str, port: int) -> Optional[dict]:
    """Modern Server List Ping (1.7+). Возвращает payload или None.

    Handshake + Status Request; ответ — JSON: players.online/max,
    description, version.name. Некоторые серверы (в т.ч. Aternos) молчат.
    """
    try:
        reader, writer = await asyncio.wait_for(
            asyncio.open_connection(host, int(port)), timeout=MODERN_SLP_TIMEOUT
        )
        try:
            handshake = bytes([0x00, 0x04]) + b"\x09localhost" + struct.pack(">H", port) + bytes([0x01])
            writer.write(struct.pack(">I", len(handshake)) + handshake)
            writer.write(b"\x01\x00")
            await writer.drain()

            length = await asyncio.wait_for(reader.read(5), timeout=MODERN_SLP_TIMEOUT)
            if not length or length[0] != 0x01:
                logger.info("modern SLP %s:%s: неверный ответ %r", host, port, length[:4])
                return None
            varint_len = 0
            shift = 0
            i = 1
            while i < len(length) and i < 5:
                varint_len |= (length[i] & 0x7F) << shift
                if not (length[i] & 0x80):
                    break
                shift += 7
                i += 1
            payload = b""
            while len(payload) < varint_len:
                chunk = await asyncio.wait_for(reader.read(varint_len - len(payload)), timeout=MODERN_SLP_TIMEOUT)
                if not chunk:
                    break
                payload += chunk
        finally:
            writer.close()
            try:
                await writer.wait_closed()
            except Exception:
                pass

        data = json.loads(payload.decode("utf-8", errors="replace"))
        players = data.get("players") or {}
        description = data.get("description") or {}
        motd = description.get("text") if isinstance(description, dict) else str(description)
        if isinstance(motd, list):
            motd = " ".join(str(p.get("text", "")) for p in motd if isinstance(p, dict))
        version = (data.get("version") or {}).get("name") or "Неизвестно"
        return {
            "ip": f"{host}:{port}",
            "is_online": True,
            "players_online": int(players.get("online") or 0),
            "players_max": int(players.get("max") or 0),
            "player_names": [str(p.get("name", "")) for p in players.get("sample", []) if isinstance(p, dict) and p.get("name")],
            "version": str(version),
        }
    except (asyncio.TimeoutError, OSError, ConnectionError, ValueError, json.JSONDecodeError) as exc:
        logger.info("modern SLP %s:%s не ответил: %s", host, port, exc)
        return None
    except Exception as exc:
        logger.warning("modern SLP %s:%s: ошибка: %s", host, port, exc)
        return None


async def _legacy_ping(host: str, port: int) -> Optional[dict]:
    """Legacy Server List Ping (0xFE 0x01). Возвращает payload или None.

    Ответ (1.7+): 0xFF 0x00 [len] [поля через \\x00]:
      MOTD | PROTOCOL | VERSION | MOTD2 | PLAYERS_ONLINE | MAX_PLAYERS
    Иногда встречаются пустые поля (двойные разделители) — они фильтруются.
    """
    logger.info("legacy ping к: %s:%s", host, port)
    try:
        reader, writer = await asyncio.wait_for(
            asyncio.open_connection(host, int(port)), timeout=LEGACY_PING_TIMEOUT
        )
        try:
            writer.write(b"\xfe\x01")
            await writer.drain()
            data = await asyncio.wait_for(reader.readexactly(3), timeout=LEGACY_PING_TIMEOUT)
            if data[:2] != b"\xff\x00":
                logger.warning("legacy ping %s:%s: неверный ответ %r", host, port, data)
                return None
            remaining = await asyncio.wait_for(
                reader.read(data[2] * 2), timeout=LEGACY_PING_TIMEOUT
            )
        finally:
            writer.close()
            try:
                await writer.wait_closed()
            except Exception:
                pass

        text = remaining.decode("utf-16-be", errors="replace")
        fields = [f.strip() for f in text.split("\x00") if f.strip()]
        if len(fields) < 3:
            logger.warning("legacy ping %s:%s: битый ответ: %r", host, port, text[:120])
            return None

        def _to_int(value: str) -> int:
            try:
                return int(value)
            except (TypeError, ValueError):
                return 0

        # Расширенный формат: [motd, protocol, version, motd2, online, max].
        if len(fields) >= 6 and _to_int(fields[1]) > 0 and fields[2].isdigit() is False:
            players_online = _to_int(fields[4])
            players_max = _to_int(fields[5])
            version = fields[2]
        else:
            # Базовый формат: [motd, online, max].
            players_online = _to_int(fields[1])
            players_max = _to_int(fields[2])
            version = "Неизвестно"

        return {
            "ip": f"{host}:{port}",
            "is_online": True,
            "players_online": players_online,
            "players_max": players_max,
            "player_names": [],
            "version": version,
        }
    except (asyncio.TimeoutError, OSError, ConnectionError, asyncio.IncompleteReadError) as exc:
        logger.info("legacy ping %s:%s не ответил: %s", host, port, exc)
        return None
    except Exception as exc:
        logger.warning("legacy ping %s:%s: ошибка: %s", host, port, exc)
        return None


async def _api_status(server_ip: str) -> Optional[dict]:
    """Статус через mcsrvstat.us API; None при недоступности API."""
    url = API_URL.format(ip=server_ip)
    logger.info("mcsrvstat запрос к: %s", url)
    timeout = aiohttp.ClientTimeout(total=REQUEST_TIMEOUT)
    try:
        async with aiohttp.ClientSession(timeout=timeout) as session:
            async with session.get(url, headers={"User-Agent": HTTP_UA}) as response:
                if response.status != 200:
                    logger.warning("mcsrvstat вернул HTTP %s для %s", response.status, server_ip)
                    return None
                data = await response.json()
        logger.info("mcsrvstat: %s -> онлайн=%s", server_ip, data.get("online"))
        return _parse_payload(server_ip, data)
    except (asyncio.TimeoutError, aiohttp.ClientError, ValueError) as exc:
        logger.warning("mcsrvstat недоступен для %s: %s", server_ip, exc)
        return None


async def get_server_status(server_ip: str, port: Optional[int] = None) -> dict:
    """Возвращает статус сервера по адресу `server_ip` (+ опциональный порт).

    Порядок попыток: mcsrvstat API -> modern SLP на правильном порту ->
    legacy ping. «Правильный порт»: переданный (server_meta) -> порт из
    ответа API -> 25565. Никогда не бросает исключений.
    """
    if not server_ip:
        logger.warning("не задан адрес сервера для статуса")
        return _empty_payload("")

    host = server_ip
    explicit_port = port
    if ":" in server_ip and not server_ip.startswith("["):
        host, _, maybe_port = server_ip.rpartition(":")
        if maybe_port.isdigit():
            explicit_port = int(maybe_port)

    api_payload = await _api_status(server_ip)
    if api_payload is not None and api_payload.get("is_online"):
        return api_payload

    api_port = (api_payload or {}).get("port")
    target_port = explicit_port or api_port or 25565

    slp = await _modern_slp(host, target_port)
    if slp is not None:
        return slp

    legacy = await _legacy_ping(host, target_port)
    if legacy is not None:
        return legacy

    if api_payload is not None:
        return api_payload
    return _empty_payload(server_ip)


async def get_servers_status(servers: List[dict]) -> Dict[str, dict]:
    """Статусы нескольких серверов (параллельный опрос).

    servers — список {server_ip, port?}; результат {server_ip: status}.
    """
    if not servers:
        return {}

    async def one(srv: dict) -> dict:
        return await get_server_status(str(srv["server_ip"]), srv.get("port"))

    statuses = await asyncio.gather(*(one(s) for s in servers))
    return {s["server_ip"]: st for s, st in zip(servers, statuses)}

File: internal/mcsrvstat/mcsrvstat.go
// Package mcsrvstat — публичный статус Minecraft-серверов (порт mcsrvstat.py).
//
// Порядок источников: REST mcsrvstat.us -> modern Server List Ping ->
// legacy ping (0xFE 0x01). Никогда не бросает ошибок наружу:
// недоступность источника = offline-статус.
package mcsrvstat

import (
	"context"
	"encoding/binary"
	"encoding/json"
	"fmt"
	"io"
	"net"
	"net/http"
	"strconv"
	"strings"
	"time"
	"unicode/utf16"
)

// Status — стандартизированный статус сервера.
type Status struct {
	IP            string
	IsOnline      bool
	PlayersOnline int
	PlayersMax    int
	PlayerNames   []string
	PlayerList    []string
	Version       string
	Port          int // 0 — неизвестен
}

const (
	apiURL        = "https://api.mcsrvstat.us/3/%s"
	mcstatusURL   = "https://api.mcstatus.io/v2/status/java/%s"
	legacyPingTO  = 5 * time.Second
	modernSLPTO   = 4 * time.Second
	requestTO     = 7 * time.Second
	playersListTO = 8 * time.Second
	httpUA        = "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36" +
		" (KHTML, like Gecko) Chrome/126.0 Safari/537.36"
)

func emptyStatus(serverIP string) Status {
	return Status{
		IP:      serverIP,
		Version: "Неизвестно",
	}
}

// GetServerStatus возвращает статус сервера по адресу (ip + опциональный порт).
//
// allowLegacy=false отключает legacy-пинг: прокси Aternos отвечает на нём даже
// на оффлайн-серверах (фейковый «онлайн»), поэтому когда панель недоступна,
// legacy использовать нельзя.
func GetServerStatus(serverIP string, port int, allowLegacy bool) Status {
	if serverIP == "" {
		return emptyStatus("")
	}

	host := serverIP
	explicitPort := port
	if i := strings.LastIndex(serverIP, ":"); i > 0 && !strings.HasPrefix(serverIP, "[") {
		if p, err := strconv.Atoi(serverIP[i+1:]); err == nil {
			host = serverIP[:i]
			explicitPort = p
		}
	}

	api := apiStatus(serverIP)
	if api != nil && api.IsOnline {
		api.PlayerList = getPlayersList(host, api.Port)
		return *api
	}

	apiPort := 0
	if api != nil && api.IsOnline {
		apiPort = api.Port
	}
	targetPort := apiPort
	if targetPort == 0 {
		targetPort = explicitPort
	}
	if targetPort == 0 {
		targetPort = 25565
	}

	if slp := modernSLP(host, targetPort); slp != nil {
		slp.PlayerList = getPlayersList(host, slp.Port)
		return *slp
	}

	if allowLegacy {
		if legacy := legacyPing(host, targetPort); legacy != nil {
			legacy.PlayerList = getPlayersList(host, legacy.Port)
			return *legacy
		}
	}

	if api != nil {
		return *api
	}
	return emptyStatus(serverIP)
}

// ------------------------------------------------------------------ //
// REST API mcsrvstat.us
// ------------------------------------------------------------------ //

type mcsrvAPIResponse struct {
	Online  bool   `json:"online"`
	IP      string `json:"ip"`
	Port    int    `json:"port"`
	Version string `json:"version"`
	Players struct {
		Online int `json:"online"`
		Max    int `json:"max"`
		List   []struct {
			Name string `json:"name"`
		} `json:"list"`
	} `json:"players"`
}

func apiStatus(serverIP string) *Status {
	client := &httpClient{timeout: requestTO}
	body, err := client.get(fmt.Sprintf(apiURL, serverIP))
	if err != nil {
		return nil
	}
	var data mcsrvAPIResponse
	if err := json.Unmarshal(body, &data); err != nil {
		return nil
	}
	names := make([]string, 0, len(data.Players.List))
	for _, p := range data.Players.List {
		if p.Name != "" {
			names = append(names, p.Name)
		}
	}
	version := data.Version
	if version == "" {
		version = "Неизвестно"
	}
	return &Status{
		IP:            serverIP,
		IsOnline:      data.Online,
		PlayersOnline: data.Players.Online,
		PlayersMax:    data.Players.Max,
		PlayerNames:   names,
		Version:       version,
		Port:          data.Port,
	}
}

// GetPlayersList возвращает ники игроков через mcstatus.io (пусто при ошибке).
func GetPlayersList(serverIP string, port int) []string {
	return getPlayersList(serverIP, port)
}

func getPlayersList(serverIP string, port int) []string {
	addr := serverIP
	if port != 0 && port != 25565 {
		addr = fmt.Sprintf("%s:%d", serverIP, port)
	}
	client := &httpClient{timeout: playersListTO}
	body, err := client.get(fmt.Sprintf(mcstatusURL, addr))
	if err != nil {
		return nil
	}
	var data struct {
		Players struct {
			List []struct {
				NameClean string `json:"name_clean"`
				NameRaw   string `json:"name_raw"`
			} `json:"list"`
		} `json:"players"`
	}
	if err := json.Unmarshal(body, &data); err != nil {
		return nil
	}
	var names []string
	for _, p := range data.Players.List {
		name := p.NameClean
		if name == "" {
			name = p.NameRaw
		}
		if name == "" {
			name = "???"
		}
		names = append(names, name)
	}
	return names
}

// ------------------------------------------------------------------ //
// modern Server List Ping (1.7+)
// ------------------------------------------------------------------ //

func modernSLP(host string, port int) *Status {
	if port == 0 {
		port = 25565
	}
	addr := net.JoinHostPort(host, strconv.Itoa(port))
	conn, err := net.DialTimeout("tcp", addr, modernSLPTO)
	if err != nil {
		return nil
	}
	defer conn.Close()
	_ = conn.SetDeadline(time.Now().Add(modernSLPTO))

	// Handshake: proto=4, host, port, next=1 (status).
	handshake := []byte{0x00, 0x04}
	handshake = append(handshake, varint(len(host))...)
	handshake = append(handshake, []byte(host)...)
	handshake = binary.BigEndian.AppendUint16(handshake, uint16(port))
	handshake = append(handshake, 0x01)

	packet := append(varint(len(handshake)), handshake...)
	packet = append(packet, 0x01, 0x00) // Status Request
	if _, err := conn.Write(packet); err != nil {
		return nil
	}

	head := make([]byte, 5)
	if _, err := io.ReadFull(conn, head); err != nil {
		return nil
	}
	if head[0] != 0x01 {
		return nil
	}
	length, err := readVarint(head[1:])
	if err != nil {
		return nil
	}
	payload := make([]byte, length)
	if _, err := io.ReadFull(conn, payload); err != nil {
		return nil
	}

	var data struct {
		Version struct {
			Name string `json:"name"`
		} `json:"version"`
		Players struct {
			Online int `json:"online"`
			Max    int `json:"max"`
			Sample []struct {
				Name string `json:"name"`
			} `json:"sample"`
		} `json:"players"`
	}
	if err := json.Unmarshal(payload, &data); err != nil {
		return nil
	}
	names := make([]string, 0, len(data.Players.Sample))
	for _, p := range data.Players.Sample {
		if p.Name != "" {
			names = append(names, p.Name)
		}
	}
	version := data.Version.Name
	if version == "" {
		version = "Неизвестно"
	}
	return &Status{
		IP:            fmt.Sprintf("%s:%d", host, port),
		Port:          port,
		IsOnline:      true,
		PlayersOnline: data.Players.Online,
		PlayersMax:    data.Players.Max,
		PlayerNames:   names,
		Version:       version,
	}
}

// ------------------------------------------------------------------ //
// legacy Server List Ping (0xFE 0x01)
// ------------------------------------------------------------------ //

func legacyPing(host string, port int) *Status {
	if port == 0 {
		port = 25565
	}
	addr := net.JoinHostPort(host, strconv.Itoa(port))
	conn, err := net.DialTimeout("tcp", addr, legacyPingTO)
	if err != nil {
		return nil
	}
	defer conn.Close()
	_ = conn.SetDeadline(time.Now().Add(legacyPingTO))

	if _, err := conn.Write([]byte{0xfe, 0x01}); err != nil {
		return nil
	}
	head := make([]byte, 3)
	if _, err := io.ReadFull(conn, head); err != nil {
		return nil
	}
	if head[0] != 0xff || head[1] != 0x00 {
		return nil
	}
	remaining := make([]byte, int(head[2])*2)
	if _, err := io.ReadFull(conn, remaining); err != nil {
		return nil
	}

	u16 := make([]uint16, 0, len(remaining)/2)
	for i := 0; i+1 < len(remaining); i += 2 {
		u16 = append(u16, binary.BigEndian.Uint16(remaining[i:i+2]))
	}
	text := string(utf16.Decode(u16))

	var fields []string
	for _, f := range strings.Split(text, "\x00") {
		if f = strings.TrimSpace(f); f != "" {
			fields = append(fields, f)
		}
	}
	if len(fields) < 3 {
		return nil
	}

	toInt := func(s string) int {
		n, err := strconv.Atoi(s)
		if err != nil || n < 0 {
			return 0
		}
		return n
	}

	var online, max int
	version := "Неизвестно"
	if len(fields) >= 6 && toInt(fields[1]) > 0 && !isAllDigits(fields[2]) {
		online = toInt(fields[4])
		max = toInt(fields[5])
		version = fields[2]
	} else {
		online = toInt(fields[1])
		max = toInt(fields[2])
	}
	return &Status{
		IP:            fmt.Sprintf("%s:%d", host, port),
		Port:          port,
		IsOnline:      true,
		PlayersOnline: online,
		PlayersMax:    max,
		Version:       version,
	}
}

func isAllDigits(s string) bool {
	if s == "" {
		return false
	}
	for _, r := range s {
		if r < '0' || r > '9' {
			return false
		}
	}
	return true
}

// ------------------------------------------------------------------ //
// вспомогательное
// ------------------------------------------------------------------ //

// varint кодирует число в протокольный формат Minecraft.
func varint(v int) []byte {
	var out []byte
	for {
		b := byte(v & 0x7F)
		v >>= 7
		if v != 0 {
			b |= 0x80
		}
		out = append(out, b)
		if v == 0 {
			return out
		}
	}
}

func readVarint(data []byte) (int, error) {
	var result int
	shift := 0
	for i, b := range data {
		result |= int(b&0x7F) << shift
		if b&0x80 == 0 {
			return result, nil
		}
		shift += 7
		if i >= 4 {
			return 0, fmt.Errorf("varint слишком длинный")
		}
	}
	return 0, fmt.Errorf("varint не завершён")
}

// httpClient — тонкая обёртка net/http с таймаутом.
type httpClient struct {
	timeout time.Duration
	client  *http.Client
}

func (c *httpClient) get(url string) ([]byte, error) {
	ctx, cancel := context.WithTimeout(context.Background(), c.timeout)
	defer cancel()

	if c.client == nil {
		c.client = &http.Client{Timeout: c.timeout}
	}
	req, err := http.NewRequestWithContext(ctx, http.MethodGet, url, nil)
	if err != nil {
		return nil, err
	}
	req.Header.Set("User-Agent", httpUA)
	resp, err := c.client.Do(req)
	if err != nil {
		return nil, err
	}
	defer resp.Body.Close()
	if resp.StatusCode != 200 {
		return nil, fmt.Errorf("HTTP %d", resp.StatusCode)
	}
	return io.ReadAll(io.LimitReader(resp.Body, 4<<20))
}
